designs/design_week2.md

```diff
@@ -0,0 +1,104 @@
+    pointer : int = 0       # A number tracking the read/write head of the file
+    filelength : int = 0    # A number tracking the file length
+
+    def __init__(self : File):
+        pass
+    
+    def read(self : File) -> int:
+        # Checks if the file is opened and in read mode
+        # Reads one i32 from file `self.fd` with the JS function
+        # (note that the current read always read from position 0)
+    
+    def write(self : File, s : int) -> int:
+        # Check if the file is opened and in write mode
+        # Write one i32 to the file `self.fd` with the JS function
+        # (note that the current write always write from position 0)
+
+    def tell(self : File) -> int:
+        # Return the current read pointer in the file
+
+    def seek(self : File, pos : int):
+        # Move the read/write pointer to `pos`
+        
+    def close(self : File):
+        # Check if the file is already closed.
+        # If not, close the file on the JS side.
+```
+
+Due to the limited time, the current implementation does not support reading from and writing to the position in the file specified by `self.pointer`. We plan the add this in the future.
+
+Additionally, we also append an implementation of the python built-in `open()` function, which returns a `File` object.
+
+```python=
+def open(mode : int) -> File:
+    newFile : File = None
+    newFile = File()
+    
+    newFile.fd = jsopen(mode)
+    newFile.closed = False
+    newFile.mode = mode
+    
+    return newFile
+```
+
+### 4. The Javascript Side of Things
+
+In the JS side, the functions calls the fileSystem APIs implemented by the BrowserFS class. Since currently the compiler does not support strings and lists, we made to choice to fix all the function to act on only one file: `test.txt`. The flags that specifies the file access mode are also limited to numbers (0 = read, 1 = write).
+
+For similar reasons, the functions only allows reading and writing of a single number at the moment. In the future, as the string and list classes are implemented, we will gradually implement more sophisticated functionalities.
+
+
+## II. Running in the IDE and Testing
+
+Run `npm install` should install the required `browserfs` extension. And then you can run `npm run build-web` to build. After serving the website, you should bw able to see class definition code and some initialization code automatically appear in the user-code text area. After clicking Run, it writes out a `test.txt` file that has '1234' in it. Then in the REPL, you can type:
+
+```
+f = open(0)
+```
+
+which opens `test.txt` in read mode.
+Then call:
+
+```
+f.read()
+```
+
+It should return the first chracter `1`.
+
+If you call `f.read()` multiple times, it should return 2, 3 and 4.
+
+Finally, you should call `f.close()`. Operating on a closed file or operating in the wrong mode will both result in errors. For now, we throw them by printing out special numbers like `99999`.
+
```

Comment:
What should be the behavior if we `read` a closed `file`? do we need a test for that?